Clustering/GMM/src/main/java/org/tribuo/clustering/gmm/GaussianMixtureModel.java

```diff
@@ -0,0 +1,394 @@
+     * for use in downstream processing, users need to map the indices using
+     * Tribuo's internal ids themselves.
+     * @return The covariances.
+     */
+    public Tensor[] getCovariances() {
+        Tensor[] copies = new Tensor[covarianceMatrices.length];
+
+        for (int i = 0; i < copies.length; i++) {
+            copies[i] = covarianceMatrices[i].copy();
+        }
+
+        return copies;
+    }
+
+    /**
+     * Returns a list of features, one per centroid.
+     * <p>
+     * This should be used in preference to {@link #getMeanVectors()}
+     * as it performs the mapping from Tribuo's internal feature ids to
+     * the externally visible feature names.
+     * </p>
+     * @return A list containing all the centroids.
+     */
+    public List<List<Feature>> getMeans() {
+        List<List<Feature>> output = new ArrayList<>(meanVectors.length);
+
+        for (int i = 0; i < meanVectors.length; i++) {
+            List<Feature> features = new ArrayList<>(featureIDMap.size());
+
+            for (VectorTuple v : meanVectors[i]) {
+                Feature f = new Feature(featureIDMap.get(v.index).getName(),v.value);
+                features.add(f);
+            }
+
+            output.add(features);
+        }
+
+        return output;
+    }
+
+    @Override
+    public Prediction<ClusterID> predict(Example<ClusterID> example) {
+        SGDVector vector;
+        if (example.size() == featureIDMap.size()) {
+            vector = DenseVector.createDenseVector(example, featureIDMap, false);
+        } else {
+            vector = SparseVector.createSparseVector(example, featureIDMap, false);
+        }
+        if (vector.numActiveElements() == 0) {
+            throw new IllegalArgumentException("No features found in Example " + example.toString());
+        }
+
+        // generate cluster responsibilities and normalize into a distribution
+        DenseVector responsibilities = new DenseVector(distributions.length);
+
+        // compute log probs
+        for (int i = 0; i < distributions.length; i++) {
+            responsibilities.set(i, distributions[i].logProbability(vector));
+        }
+
+        // add mixing distribution
+        responsibilities.intersectAndAddInPlace(mixingDistribution, Math::log);
+
+        // convert from log space into probabilities
+        double sum = responsibilities.logSumExp();
+        responsibilities.scalarAddInPlace(-sum);
+        responsibilities.foreachInPlace(Math::exp);
+
+        // compute output prediction
+        ClusterID max = null;
+        Map<String, ClusterID> scores = new HashMap<>();
+        for (int i = 0; i < distributions.length; i++) {
+            ClusterID tmp = new ClusterID(i, responsibilities.get(i));
+            if (max == null || tmp.getScore() > max.getScore()) {
+                max = tmp;
+            }
+        }
+
+        return new Prediction<>(max,scores,vector.size(),example,generatesProbabilities);
```

Comment:
does it make sense for `generatesProbabilities` to ever not be `true` here?

Reply:
It's set to true in the constructor and the field is final, so it can't ever be false. I can change it to be a literal true if you want.

Reply:
I think I'd prefer that slightly, or at least a comment to that effect.